### PR TITLE
fixes #17354 - test and improve idempotency of OS fact parser

### PR DESCRIPTION
**Pull request by domcleal**
Contains the commit from #4025 which fixes an issue when parsing OSes with only a major version.

### CODE REVIEW

File: app/models/operatingsystems/aix.rb [hunk 1]
```diff
@@ -1,9 +1,8 @@
 class AIX < Operatingsystem
   PXEFILES = {:kernel => "powerpc", :initrd => "initrd"}
 
-  # Override the class representation, as this breaks many rails helpers
-  def class
-    Operatingsystem
+  class << self
```
Comment (timogoebel) [inline app/models/operatingsystems/aix.rb]: You could add this to the superclass to not repeat yourself:

``` ruby
def self.inherited(child)
  child.instance_eval do
    # rubocop:disable Rails/Delegate
    def model_name
      superclass.model_name
      end
    # rubocop:enable Rails/Delegate
  end
  super
end
```

Comment (domcleal) [reply app/models/operatingsystems/aix.rb]: Thanks, I've used this and removed existing model_name impls.

File: app/services/puppet_fact_parser.rb [hunk 2]
```diff
@@ -40,14 +40,20 @@ def operatingsystem
...
       elsif facts[:lsbdistdescription]
         family = os.deduce_family || 'Operatingsystem'
         os.description = family.constantize.shorten_description facts[:lsbdistdescription]
       end
     end
-    os.save!
-    os
+
+    if os.new_record?
+      os.save!
```
Comment (dLobatog) [inline app/services/puppet_fact_parser.rb:51]: nitpick, why checking whether it's a new record? I suppose already existing Operatingsystem instances can be reloaded just the same by calling `.find` and you can avoid the entire if condition just with:

``` ruby
os.save!
OperatingSystem.find(os.id)
```

:question:

File: app/services/puppet_fact_parser.rb [hunk 3]
```diff
@@ -40,14 +40,20 @@ def operatingsystem
     end
     if os.description.blank?
       if os_name == 'SLES'
-        os.description = os_name + ' ' + orel.gsub!('.', ' SP')
+        os.description = os_name + ' ' + orel.gsub('.', ' SP')
```
Comment (dLobatog) [inline app/services/puppet_fact_parser.rb:43]: I'm not sure I understand why this change. It doesn't seem to change the behavior as we're not using `orel` anywhere after this anyway. gsub! (in place) should be slightly faster. If we're just going to make one substitution `orel['.'] = ' SP'` [should be even faster](https://github.com/JuanitoFatas/fast-ruby#stringsub-vs-stringgsub-vs-string-code) (petty, I know)

**Review by dLobatog — CHANGES_REQUESTED** [review]
Just 2 minor comments on the parser, I tried making those changes and tests passed just fine. Thanks @domcleal

File: app/services/puppet_fact_parser.rb (hunk 2 above)
Comment (domcleal) [reply app/services/puppet_fact_parser.rb:51]: I was trying to avoid doing the extra .find call if the OS already existed. If it's a new record then its class will be Operatingsystem with `family` set to say, `Debian` so has to be reloaded to be the right class.

If the OS already existed and we're just updating it, the family will already be known, it'll be a `Debian` class and so saving it is enough - it doesn't need to reload from the DB, saving a call.

File: app/services/puppet_fact_parser.rb (hunk 3 above)
Comment (domcleal) [reply app/services/puppet_fact_parser.rb:43]: Yeah, I found this in the SLES parser test before adding `.dup` to `os_release`. They fix the same issue, but I left this here because it's more correct.

`#gsub!` returns `nil` if there are no changes made, so shouldn't be used in concatenation:

``` ruby
2.3.0 :001 > "1.2".gsub!('.', ' SP')
 => "1 SP2" 
2.3.0 :002 > "1.2".gsub!('.', ' SP').gsub!('.', ' SP')
 => nil 
```

What was happening in the tests was the same facts hash was parsed twice and using `gsub!` modified the original fact. When the test called the parser again, the gsub! returned nil and the test failed trying to concatenate String and NilClass.

**Review by dLobatog — APPROVED** [review]

**Comment by dLobatog** [comment]
Thanks @domcleal !

**Comment by dLobatog** [comment]
Thanks @domcleal !
